tests/test_butlerstd.py

```diff
@@ -0,0 +1,346 @@
+        self.registry = Registry()
+        self.current_ref = None
+        self.mockImages = mock_images_f
+
+    def getURI(self, ref, collections=None):
+        mocked = mock.Mock(name="ButlerURI")
+        mocked.geturl.return_value = f"file:/{self.datastore.root}"
+        return mocked
+
+    def getDataset(self, datid):
+        return self.get(datid)
+
+    def get(self, ref, collections=None):
+        # Butler.get gets a DatasetRef, but can take an DatasetRef or DatasetId
+        # DatasetId is type alias for UUID's, which are hex-strings when
+        # serialized. We short it to an integer, because We use an integer to
+        # read a particular file in FitsFactory. This means we got to cast
+        # all these different objects to int (somehow). Firstly, it's one of
+        # our mocks, dig out the value we really care about:
+        if isinstance(ref, (DatasetId, DatasetRef)):
+            ref = ref.ref
+
+        # that value can be an int, a simple str(int) (used in testing only),
+        # a large hex UUID string, or a UUID object. Duck-type them to int
+        if isinstance(ref, uuid.UUID):
+            ref = ref.int
+        elif isinstance(ref, str):
+            try:
+                ref = uuid.UUID(ref).int
+            except (ValueError, AttributeError):
+                # likely a str(int)
+                pass
+
+        # Cast to int to cover for all eventualities
+        ref = int(ref)
+        self.current_ref = ref
+
+        # Finally we can proceed with mocking. Butler.get (the way we use it at
+        # least) returns an Exposure[F/I/...] object. Exposure is like our
+        # LayeredImage. We need to mock every attr, method and property that we
+        # call the standardizer. We shortcut the results to match the KBMODV1.
+        hdul = FitsFactory.get_fits(ref % FitsFactory.n_files, spoof_data=True)
+        prim = hdul["PRIMARY"].header
+
+        mocked = mock.Mock(
+            name="Exposure",
+            spec_set=[
+                "visitInfo",
+                "info",
+                "hasWcs",
+                "getWidth",
+                "getHeight",
+                "getFilter",
+                "image",
+                "variance",
+                "mask",
+                "wcs",
+            ],
+        )
+
+        # General metadata mocks
+        mocked.visitInfo.date.toAstropy.return_value = Time(hdul["PRIMARY"].header["DATE-AVG"], format="isot")
+        mocked.info.id = prim["EXPID"]
+        mocked.getWidth.return_value = hdul[1].header["NAXIS1"]
+        mocked.getHeight.return_value = hdul[1].header["NAXIS2"]
+        mocked.info.getFilter().physicalLabel = prim["FILTER"]
+
+        # Rubin Sci. Pipes. return their own internal SkyWcs object. We mock a
+        # Header that'll work with ButlerStd instead. It works because in the
+        # STD we cast SkyWcs to dict-like thing, from which we make a WCS. What
+        # happens if SkyWcs changes though?
```

Comment:
We could add a test that creates a SkyWcs and makes sure it can convert to a dict-like thing? Or, more realistically we can do a fix if it changes later.

Reply:
`SkyWcs` comes from the [`lsst.afw`](https://github.com/lsst/afw), if I mock it I can mock it in a way that it dictifies, but that still doesn't tell me anything about the real `SkyWcs` from `AFW` C++ code used by Rubin. Thankfully this is not likely to change, but generally doing these tests is a pain. We need a canary CI that will build from Rubin weeklies, but the setup is hard so it's a big job to do and the CI run itself will probably go on for a few hours as it would need to build the weekly, run some data through it and then run the KBMOD tests.

This has been my long-time goal for a while though.

Reply:
I'm also not sure how to create a `SkyWcs` either. I was just suggesting that, if we could create one, that we could use that to watch for breakages.